src/storage/files.cpp

```diff
@@ -41,126 +42,124 @@ using blackhole::attribute_list;
 files_t::files_t(context_t& context, const std::string& name, const dynamic_t& args):
     category_type(context, name, args),
     m_log(context.log(name)),
-    m_parent_path(args.as_object().at("path").as_string())
+    m_parent_path(args.as_object().at("path").as_string()),
+    io_loop(),
+    io_work(io_loop),
+    thread([&](){
+        io_loop.run();
+    })
 { }
 
 files_t::~files_t() {
-    // Empty.
+    io_work = boost::none;
+    thread.join();
 }
 
 void
-files_t::read(callback<std::string> cb, const std::string& collection, const std::string& key) {
-    std::lock_guard<std::mutex> guard(m_mutex);
-
-    const fs::path file_path(m_parent_path / collection / key);
+files_t::read(const std::string& collection, const std::string& key, callback<std::string> cb) {
+    io_loop.post([=](){
+        const fs::path file_path(m_parent_path / collection / key);
 
-    if(!fs::exists(file_path)) {
-        return cb(make_exceptional_future<std::string>(std::make_error_code(std::errc::no_such_file_or_directory),
-            file_path.string()
-        ));
-    }
+        if(!fs::exists(file_path)) {
+            auto ec = std::make_error_code(std::errc::no_such_file_or_directory);
```

Comment:
Look, domain and asynchronous logic intersection here. Isn't it better to decompose file sync reading/writing into separate functions, that throw exceptions (like it was before) and then make them private and provide async overloads, which does only:

```c++
loop.post([&]() {
    try { 
        cb(read(...)); 
    catch (...) {
        cb(make_exceptional_future<std::string>(std::current_exception());
    }
})
```

?